spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindImproperSynchronization.java

```diff
@@ -0,0 +1,803 @@
+                                lock, backingCollection, exposingMethodsMessage));
+            }
+        }
+    }
+
+    private void reportInheritedBackingCollectionBug(XField lock, MultiMap<XField, BugInfo> possibleBugs,
+            XField backingCollection) {
+        for (BugInfo bugInfo : possibleBugs.get(lock)) {
+            bugReporter.reportBug(bugInfo
+                    .createBugInstance(this,
+                            "USBC_UNSAFE_SYNCHRONIZATION_WITH_INHERITABLE_BACKING_COLLECTION",
+                            lock, backingCollection));
+        }
+    }
+
+    private void reportBadCollectionObjectBugs(XField lock, MultiMap<XField, BugInfo> possibleBugs,
+            XField backingCollection) {
+        for (BugInfo bugInfo : possibleBugs.get(lock)) {
+            bugReporter.reportBug(bugInfo
+                    .createBugInstance(this,
+                            "USBC_UNSAFE_SYNCHRONIZATION_WITH_BACKING_COLLECTION",
+                            lock, backingCollection));
+        }
+    }
+
+    private void clearState() {
+        exposingMethods.clear();
+        synchronizedMethods.clear();
+        declaredLockObjects.clear();
+        inheritedLockObjects.clear();
+        declaredLockAccessors.clear();
+        lockAccessorsInHierarchy.clear();
+        lockUsingMethodsInHierarchy.clear();
+        declaredInheritedFieldAccessingMethods.clear();
+        potentialObjectBugContainingMethods.clear();
+        potentialInheritedBugContainingMethods.clear();
+
+        // Backing collections
+        declaredCollectionLockObjects.clear();
+        inheritedCollectionLockObjects.clear();
+        declaredCollectionAccessors.clear();
+        collectionAccessorsInHierarchy.clear();
+        backingCollections.clear();
+        badBackingCollections.clear();
+        inheritableBackingCollections.clear();
+    }
+
+    private static class BugInfo {
+        private final XMethod inMethod;
+        private final SourceLineAnnotation sourceLineAnnotation;
+
+        private BugInfo(BytecodeScanningDetector detector) {
+            this.inMethod = detector.getXMethod();
+            this.sourceLineAnnotation = SourceLineAnnotation.fromVisitedInstruction(detector);
+        }
+
+        private BugInstance createBugInstance(BytecodeScanningDetector detector, String bugType, XField lock, XField assignedBackingCollection,
+                String... extraMessages) {
+            BugInstance bugInstance = new BugInstance(detector, bugType, NORMAL_PRIORITY)
+                    .addClassAndMethod(inMethod)
+                    .addField(lock)
+                    .addReferencedField(FieldAnnotation.fromXField(assignedBackingCollection))
+                    .addSourceLine(sourceLineAnnotation);
+            for (String message : extraMessages) {
+                bugInstance.addString(message);
+            }
+
+            return bugInstance;
+        }
+    }
```

Comment:
As far as I can see, this function is called three times, either with nothing as `extraMessages`, or with one item.
IMO it would make a code slightly easier to read, if it would look like something like this:

```suggestion
        private BugInstance createBugInstance(BytecodeScanningDetector detector, String bugType, XField lock, XField assignedBackingCollection) {
            return new BugInstance(detector, bugType, NORMAL_PRIORITY)
                    .addClassAndMethod(inMethod)
                    .addField(lock)
                    .addReferencedField(FieldAnnotation.fromXField(assignedBackingCollection))
                    .addSourceLine(sourceLineAnnotation);
        }
    }
    
    private BugInstance createBugInstance(BytecodeScanningDetector detector, String bugType, XField lock, XField assignedBackingCollection, String extraMessage) {
            BugInstance bugInstance = createBugInstance(detector, bugType, lock, assignedBackingCollection);
            bugInstance.addString(extraMessage);

            return bugInstance;
        }
    }
```